.github/workflows/build.yml

```diff
@@ -0,0 +1,180 @@
+    - name: Verify
+      run: make verify
+
+  test:
+    name: Unit tests
+    runs-on: ubuntu-latest
+    env:
+      DOCKER_HOSTNAME: "localhost:5000"
+      DOCKER_NAMESPACE: m4d-system
+    steps:
+    - name: Set up JDK 11
+      uses: actions/setup-java@v1
+      with:
+        java-version: 11
+    - name: Install Go
+      uses: actions/setup-go@v2
+      with:
+        go-version: 1.13
+    - uses: actions/checkout@v2
+    - name: Cache dependencies
+      uses: actions/cache@v2
+      with:
+        path: |
+          ~/.m2/repository
+          ~/go/pkg/mod
+          ${{ github.workspace }}/hack/tools/bin
+        key: ${{ runner.os }}-ut-m2-${{ hashFiles('**/pom.xml') }}-go-${{ hashFiles('**/go.sum') }}
+        restore-keys: ${{ runner.os }}-ut-m2
+    - name: Test
+      run: make test
+
+  integration-tests:
+    name: Integration Tests
+    runs-on: ubuntu-latest
+    env:
+      DOCKER_HOSTNAME: "localhost:5000"
+      DOCKER_NAMESPACE: m4d-system
+    steps:
+    - name: Set up JDK 11
+      uses: actions/setup-java@v1
+      with:
+        java-version: 11
+    - name: Install Go
+      uses: actions/setup-go@v2
+      with:
+        go-version: 1.13
+    - uses: actions/checkout@v2
+    - name: Cache dependencies
+      uses: actions/cache@v2
+      with:
+        path: |
+          ~/.m2/repository
+          ~/go/pkg/mod
+          ${{ github.workspace }}/hack/tools/bin
+        key: ${{ runner.os }}-it-m2-${{ hashFiles('**/pom.xml') }}-go-${{ hashFiles('**/go.sum') }}
+        restore-keys: ${{ runner.os }}-it-m2
+    - name: Install tools
+      run: make install-tools
+    - name: Integration tests
+      run: make run-integration-tests
+    - run: docker images
+    - name: Save images
+      if: ${{ github.event_name != 'pull_request' && env.ACTIVE_CI == 'gh_actions' }}
+      run: make save-images
+    - name: Upload images
+      if: ${{ github.event_name != 'pull_request' && env.ACTIVE_CI == 'gh_actions' }}
+      uses: actions/upload-artifact@v2
+      with:
+        name: images
+        path: images.tar
+        retention-days: 1
+
+  push_images:
+    name: Push images
+    runs-on: ubuntu-latest
+    env:
+      DOCKER_HOSTNAME: "localhost:5000"
+      DOCKER_NAMESPACE: m4d-system
+    needs: [verify, test, integration-tests]
+    if: ${{ github.event_name != 'pull_request' }}
+    steps:
+    - uses: actions/checkout@v2
+    - name: Download artifact
+      if: ${{ github.event_name != 'pull_request' && env.ACTIVE_CI == 'gh_actions' }}
+      uses: actions/download-artifact@v2
+      with:
+        name: images
+    - name: Load images
+      if: ${{ github.event_name != 'pull_request' && env.ACTIVE_CI == 'gh_actions' }}
+      run: docker load -i images.tar
+    - run: docker images
+    - name: Publish images
+      if: ${{ github.event_name != 'pull_request' && env.ACTIVE_CI == 'gh_actions' }}
+      run: sudo chmod o+x /etc/docker && make docker-retag-and-push-public && make helm-push-public
```

Comment:
[FUTURE] 
I think that this eventually be a separate workflow that runs on master and tags? Like build tag and push (leverage layer caching for performance), all in a workflow that runs in push to master or a `v*` tag.

Reply:
Yes. Separate workflows might be better here. I wanted to keep one for now to avoid duplication for now. If we find some templating mechanism that makes sure that both workflows always run the tests the same way I think we should split them.